{"cells":[{"metadata":{"trusted":true},"cell_type":"code","source":"from itertools import islice\n\nimport numpy as np\nimport pandas as pd\nfrom sklearn.metrics import f1_score\nimport csv \n\n\nimport tensorflow as tf\nfrom keras.preprocessing import text, sequence\nfrom keras.utils import to_categorical\nfrom keras.preprocessing.sequence import pad_sequences\n#from keras import backend as K\n\n#from tensorflow.keras.layers import Input, Embedding, Bidirectional, Dense, LSTM, TimeDistributed, Lambda, SpatialDropout1D\n#from tensorflow.keras.layers import Conv1D, GlobalMaxPooling1D, GlobalAveragePooling1D, concatenate, SpatialDropout1D\n\n#from tensorflow.keras.models import Model\n#from keras.optimizers import SGD, Adam, RMSprop","execution_count":1,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"semgroups = \"/kaggle/input/thesis/SemGroups_2018.txt\"\nsemtypes = \"/kaggle/input/thesis/SemanticTypes_2018AB.txt\"\nmedmentions = \"/kaggle/input/thesis/corpus_pubtator.txt\"\numls_concept = \"/kaggle/input/thesis/MRCONSO.RRF\"\ntrain_docs_ids = \"../input/doc-ids/corpus_pubtator_pmids_trng.txt\"\nval_docs_ids = \"../input/doc-ids/corpus_pubtator_pmids_dev.txt\"\ntest_docs_ids = \"../input/doc-ids/corpus_pubtator_pmids_test.txt\"","execution_count":2,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def read_in_chunks(file_object, n=10000):\n    \"\"\"Lazy function (generator) to read a file piece by piece.\n    Default chunk size: 1k.\"\"\"\n    while True:\n        data = list(islice(file_object, n))\n        if not data:\n            break\n        yield data\n        \ndef umls_concepts():\n    umls_concepts = {}\n    with open(umls_concept) as f:\n        for piece in read_in_chunks(f):\n            for line in piece:\n                if line != \"\":\n                    line_list = line.split(\"|\")\n                    if line_list[1] == 'ENG':\n                        umls_concepts[line_list[0]] = line_list[14]\n    return umls_concepts\n\ndef umls_semtype():\n    umls_semtype = {}\n    with open(semgroups) as f:\n        f = f.read().split(\"\\n\")\n        for line in f:\n            if line != \"\":\n                line_list = line.split(\"|\")\n                umls_semtype[line_list[2]] = [line_list[1], line_list[3]]\n    return umls_semtype","execution_count":3,"outputs":[]},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"umls_concepts = umls_concepts()\numls_semtype = umls_semtype()\n\numls_semtype['T131']","execution_count":4,"outputs":[{"output_type":"execute_result","execution_count":4,"data":{"text/plain":"['Chemicals & Drugs', 'Hazardous or Poisonous Substance']"},"metadata":{}}]},{"metadata":{"trusted":true},"cell_type":"code","source":"len(umls_concepts), len(umls_semtype)","execution_count":5,"outputs":[{"output_type":"execute_result","execution_count":5,"data":{"text/plain":"(4412440, 127)"},"metadata":{}}]},{"metadata":{},"cell_type":"markdown","source":"## Build Data"},{"metadata":{"trusted":true},"cell_type":"code","source":"%%capture\n!pip install git+https://github.com/ArshSekhon/pubtator_loader.git\n\nfrom pubtator_loader import PubTatorCorpusReader\ndataset_reader = PubTatorCorpusReader(medmentions)\n\ncorpus = dataset_reader.load_corpus() ","execution_count":11,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def build_data():\n    sequences = []\n    labels = []\n    articles_ids = []\n    entire_corpus = []\n    full_maps = []\n    for doc in corpus:\n        row = []\n        umls_map = []\n        full_text = doc.get_space_separated_title_and_abstract()\n        entire_corpus.append(full_text)\n        full_text_replaced = full_text\n        for i, ent in enumerate(doc.entities):\n            umls_concept_code = ent.entity_id\n            text_segment = ent.text_segment\n            full_text_replaced = full_text_replaced.replace(text_segment, \" \"+umls_concept_code+\" \", 1)\n            umls_map.append([text_segment, umls_concept_code])\n            \n        full_maps.append(umls_map)\n        full_text_list = full_text_replaced.split(\" \")        \n        entities_list = [ent.entity_id for ent in doc.entities]\n        articles_ids.append(doc.id)\n        \n        for word in full_text_list:\n            if word != '':\n                idx = -1\n                indices = [i for i, x in enumerate(entities_list) if x == word]\n                if word in entities_list:\n                    idx = entities_list.index(word)   \n                if idx > -1:\n                    if \",\" in doc.entities[idx].semantic_type_id:\n                        sem_ent = doc.entities[idx].semantic_type_id.split(\",\")[0]\n                    else:\n                        sem_ent = doc.entities[idx].semantic_type_id\n                    row.append([word, sem_ent])\n                else:\n                    row.append([word, 'o'])\n                    \n        sequences.append(list(np.array(row)[:,0]))\n        labels.append(list(np.array(row)[:,1]))\n        \n    original_text_list = sequences\n    original_text = []\n    for i, text in enumerate(original_text_list):\n        text_string = \"|i|\".join(text)\n        for k, v in full_maps[i]:\n            text_string = text_string.replace(v, k, 1)\n        text_string_list = text_string.split(\"|i|\")\n        original_text.append(text_string_list)\n    \n    return sequences, labels, entire_corpus, articles_ids, original_text\n    \nsequences, labels, entire_corpus, articles_ids, original_text = build_data()","execution_count":89,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"assert sequences[0][0] == \"C4308010\"\nassert labels[0][0] == \"T116\"\nassert sequences[0][-5] == \"C0854135\"\nassert labels[0][-5] == \"T047\"\n\nassert sequences[5][1] == \"C0870811\"\nassert labels[5][1] == \"T054\"\nassert sequences[5][-2] == \"C0243095\"\nassert labels[5][-2] == \"T033\"\n\ntext_0 = \"\"\"DCTN4 as a modifier of chronic Pseudomonas aeruginosa infection in cystic fibrosis Pseudomonas aeruginosa (Pa) infection in cystic fibrosis (CF) patients is associated with worse long-term pulmonary disease and shorter survival, and chronic Pa infection (CPA) is associated with reduced lung function, faster rate of lung decline, increased rates of exacerbations and shorter survival. By using exome sequencing and extreme phenotype design, it was recently shown that isoforms of dynactin 4 (DCTN4) may influence Pa infection in CF, leading to worse respiratory disease. The purpose of this study was to investigate the role of DCTN4 missense variants on Pa infection incidence, age at first Pa infection and chronic Pa infection incidence in a cohort of adult CF patients from a single centre. Polymerase chain reaction and direct sequencing were used to screen DNA samples for DCTN4 variants. A total of 121 adult CF patients from the Cochin Hospital CF centre have been included, all of them carrying two CFTR defects: 103 developed at least 1 pulmonary infection with Pa, and 68 patients of them had CPA. DCTN4 variants were identified in 24% (29/121) CF patients with Pa infection and in only 17% (3/18) CF patients with no Pa infection. Of the patients with CPA, 29% (20/68) had DCTN4 missense variants vs 23% (8/35) in patients without CPA. Interestingly, p.Tyr263Cys tend to be more frequently observed in CF patients with CPA than in patients without CPA (4/68 vs 0/35), and DCTN4 missense variants tend to be more frequent in male CF patients with CPA bearing two class II mutations than in male CF patients without CPA bearing two class II mutations (P = 0.06). Our observations reinforce that DCTN4 missense variants, especially p.Tyr263Cys, may be involved in the pathogenesis of CPA in male CF.\"\"\"\nsequences_codes_3 = [\"C0277814\", \"C0806909\", \"C0231452\", \":\", \"An\", \"C1523987\", \"to\", \"C0231472\", \"C0806909\", \"C0231452\", \"for\", \"determining\", \"presence\"]\nlabels_3 = [\"T033\", \"T081\", \"T042\", \"o\", \"o\", \"T077\", \"o\", \"T082\", \"T081\", \"T042\", \"o\", \"o\", \"o\"]\n\nassert entire_corpus[0] == text_0\nassert sequences[3][:13] == sequences_codes_3\nassert labels[3][:13] == labels_3","execution_count":93,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Data Visualization"},{"metadata":{"trusted":true},"cell_type":"code","source":"#first article\nentire_corpus[0]","execution_count":94,"outputs":[{"output_type":"execute_result","execution_count":94,"data":{"text/plain":"'DCTN4 as a modifier of chronic Pseudomonas aeruginosa infection in cystic fibrosis Pseudomonas aeruginosa (Pa) infection in cystic fibrosis (CF) patients is associated with worse long-term pulmonary disease and shorter survival, and chronic Pa infection (CPA) is associated with reduced lung function, faster rate of lung decline, increased rates of exacerbations and shorter survival. By using exome sequencing and extreme phenotype design, it was recently shown that isoforms of dynactin 4 (DCTN4) may influence Pa infection in CF, leading to worse respiratory disease. The purpose of this study was to investigate the role of DCTN4 missense variants on Pa infection incidence, age at first Pa infection and chronic Pa infection incidence in a cohort of adult CF patients from a single centre. Polymerase chain reaction and direct sequencing were used to screen DNA samples for DCTN4 variants. A total of 121 adult CF patients from the Cochin Hospital CF centre have been included, all of them carrying two CFTR defects: 103 developed at least 1 pulmonary infection with Pa, and 68 patients of them had CPA. DCTN4 variants were identified in 24% (29/121) CF patients with Pa infection and in only 17% (3/18) CF patients with no Pa infection. Of the patients with CPA, 29% (20/68) had DCTN4 missense variants vs 23% (8/35) in patients without CPA. Interestingly, p.Tyr263Cys tend to be more frequently observed in CF patients with CPA than in patients without CPA (4/68 vs 0/35), and DCTN4 missense variants tend to be more frequent in male CF patients with CPA bearing two class II mutations than in male CF patients without CPA bearing two class II mutations (P = 0.06). Our observations reinforce that DCTN4 missense variants, especially p.Tyr263Cys, may be involved in the pathogenesis of CPA in male CF.'"},"metadata":{}}]},{"metadata":{"trusted":true},"cell_type":"code","source":"def tokens_tags(data_X, data_y):\n    vocab = []\n    tags = []\n    for seq in data_X:\n        for word in np.unique(seq):\n            if word.lower() not in vocab:\n                vocab.append(word.lower())\n    for seq_ent in data_y:\n        for ent in seq_ent:\n            if ent not in tags:\n                tags.append(ent)\n                \n    return len(vocab), len(tags), max([len(s) for s in sequences])\n\nnum_tokens, num_tags, maxlen = tokens_tags(sequences, labels)\nprint('Number of unique tokens ', num_tokens)\nprint('Max length of sequence ', maxlen)\nprint('Number of unique tags ', num_tags)","execution_count":95,"outputs":[{"output_type":"stream","text":"Number of unique tokens  66317\nMax length of sequence  685\nNumber of unique tags  127\n","name":"stdout"}]},{"metadata":{"trusted":true},"cell_type":"code","source":"def minority_classes(entities):\n    class_count = {}\n    for seq_tags in entities:\n        for tag in seq_tags:\n            if tag not in class_count.keys():\n                class_count[tag] = 1\n            else:\n                class_count[tag] += 1\n    \n    threeshold = 10\n\n    less_represented_classes = {}\n    \n    for k, v in class_count.items():\n        if v < threeshold:\n            less_represented_classes[k] = v\n\n    return less_represented_classes\n\nminority_classes(labels)            ","execution_count":96,"outputs":[{"output_type":"execute_result","execution_count":96,"data":{"text/plain":"{'T021': 2, 'T127': 2}"},"metadata":{}}]},{"metadata":{},"cell_type":"markdown","source":"## Train Test Val split"},{"metadata":{"trusted":true},"cell_type":"code","source":"def train_test_val_split():\n    train_sequences, test_sequences, val_sequences = [], [], []\n    train_labels, test_labels, val_labels = [], [], []\n    train_text, test_text, val_text = [], [], []\n    train_original_sequences, test_original_sequences ,val_original_sequences = [], [], []\n    \n    with open(train_docs_ids) as train_ids:\n        train_id_list = []\n        train_ids = train_ids.read().split(\"\\n\")\n        for line in train_ids:\n            train_id_list.append(line)\n    \n    with open(val_docs_ids) as val_ids:\n        val_id_list = []\n        val_ids = val_ids.read().split(\"\\n\")\n        for line in val_ids:\n            val_id_list.append(line)\n\n    with open(test_docs_ids) as test_ids:\n        test_id_list = []\n        test_ids = test_ids.read().split(\"\\n\")\n        for line in test_ids:\n            test_id_list.append(line)\n\n\n    for x, y, txt, idx, ox in zip(sequences, labels, entire_corpus, articles_ids, original_text):\n        idx = str(idx)\n        if idx in train_id_list:\n            train_text.append(txt)\n            train_sequences.append(x)\n            train_labels.append(y)\n            train_original_sequences.append(ox)\n        elif idx in val_id_list:\n            val_text.append(txt)\n            val_sequences.append(x)\n            val_labels.append(y)\n            val_original_sequences.append(ox)\n        elif idx in test_id_list:\n            test_text.append(txt)\n            test_sequences.append(x)\n            test_labels.append(y)\n            test_original_sequences.append(ox)\n            \n    return (train_sequences, \n            train_labels, \n            train_text,\n            train_original_sequences,\n            val_sequences, \n            val_labels, \n            val_text, \n            val_original_sequences,\n            test_sequences,\n            test_labels,\n            test_text,\n            test_original_sequences)\n\n\n(train_sequences, \n train_labels, \n train_text,\n train_original_sequences,\n val_sequences, \n val_labels, \n val_text, \n val_original_sequences,\n test_sequences,\n test_labels,\n test_text,\n test_original_sequences) = train_test_val_split()","execution_count":98,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"print(len(train_sequences))\nprint(len(train_labels))\nprint(len(train_text))  \nprint(len(val_sequences))\nprint(len(val_labels))\nprint(len(val_text)) \nprint(len(test_sequences))\nprint(len(test_labels))\nprint(len(test_text))","execution_count":99,"outputs":[{"output_type":"stream","text":"2635\n2635\n2635\n878\n878\n878\n879\n879\n879\n","name":"stdout"}]},{"metadata":{"trusted":true},"cell_type":"code","source":"num_tokens_train, num_tags_train = tokens_tags(train_sequences, train_labels)[:2]\nprint('Number of unique tokens in train data ', num_tokens_train)\nprint('Number of unique tags in train data ', num_tags_train)\nnum_tokens_val, num_tags_val = tokens_tags(val_sequences, val_labels)[:2]\nprint('Number of unique tokens in val data ', num_tokens_val)\nprint('Number of unique tags in val data ', num_tags_val)\nnum_tokens_test, num_tags_test = tokens_tags(test_sequences, test_labels)[:2]\nprint('Number of unique tokens in test data ', num_tokens_test)\nprint('Number of unique tags in test data ', num_tags_test)","execution_count":100,"outputs":[{"output_type":"stream","text":"Number of unique tokens in train data  48187\nNumber of unique tags in train data  127\nNumber of unique tokens in val data  23181\nNumber of unique tags in val data  125\nNumber of unique tokens in test data  23252\nNumber of unique tags in test data  124\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"## Save all data to csv"},{"metadata":{"trusted":true},"cell_type":"code","source":"def extract_rows(text, seqs, labels, otext):\n    rows = []\n    for txt, x, ox, y in zip(text, seqs, otext, labels):\n        rows.append([txt, x, ox, y])\n    return rows\n\ntest_rows = extract_rows(test_text, test_sequences, test_labels, test_original_sequences)\ntrain_rows = extract_rows(train_text, train_sequences, train_labels, train_original_sequences)\nval_rows = extract_rows(val_text, val_sequences, val_labels, val_original_sequences)","execution_count":101,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def save_csv(rows, name):\n    # field names \n    fields = ['Full Text', 'UMLS Text', 'Original Text', 'Entities'] \n    \n    with open(name+'.csv', 'w') as f: \n        write = csv.writer(f) \n        write.writerow(fields) \n        write.writerows(rows) \n\nsave_csv(test_rows, 'test')\nsave_csv(train_rows, 'train')\nsave_csv(val_rows, 'val')","execution_count":102,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data = pd.read_csv('./train.csv')\ndata.head()","execution_count":103,"outputs":[{"output_type":"execute_result","execution_count":103,"data":{"text/plain":"                                           Full Text  \\\n0  DCTN4 as a modifier of chronic Pseudomonas aer...   \n1  Prevascularized silicon membranes for the enha...   \n2  Seated maximum flexion: An alternative to stan...   \n3  The Relationship Between Distance and Post-ope...   \n4  Promoting lifestyle behaviour change and well-...   \n\n                                           UMLS Text  \\\n0  ['C4308010', 'as', 'a', 'modifier', 'of', 'C08...   \n1  ['C0042382', 'C0037114', 'C1706182', 'for', 't...   \n2  ['C0277814', 'C0806909', 'C0231452', ':', 'An'...   \n3  ['The', 'C0439849', 'Between', 'C0012751', 'an...   \n4  ['C0033414', 'C0870811', 'and', 'C0424578', 'i...   \n\n                                       Original Text  \\\n0  ['DCTN4', 'as', 'a', 'modifier', 'of', 'chroni...   \n1  ['Prevascularized', 'silicon', 'membranes', 'f...   \n2  ['Seated', 'maximum', 'flexion', ':', 'An', 'a...   \n3  ['The', 'Relationship', 'Between', 'Distance',...   \n4  ['Promoting', 'lifestyle behaviour change', 'a...   \n\n                                            Entities  \n0  ['T116', 'o', 'o', 'o', 'o', 'T047', 'o', 'T04...  \n1  ['T169', 'T109', 'T073', 'o', 'o', 'T052', 'o'...  \n2  ['T033', 'T081', 'T042', 'o', 'o', 'T077', 'o'...  \n3  ['o', 'T080', 'o', 'T081', 'o', 'T033', 'T053'...  \n4  ['T052', 'T054', 'o', 'T033', 'o', 'T101', 'o'...  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Full Text</th>\n      <th>UMLS Text</th>\n      <th>Original Text</th>\n      <th>Entities</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>DCTN4 as a modifier of chronic Pseudomonas aer...</td>\n      <td>['C4308010', 'as', 'a', 'modifier', 'of', 'C08...</td>\n      <td>['DCTN4', 'as', 'a', 'modifier', 'of', 'chroni...</td>\n      <td>['T116', 'o', 'o', 'o', 'o', 'T047', 'o', 'T04...</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>Prevascularized silicon membranes for the enha...</td>\n      <td>['C0042382', 'C0037114', 'C1706182', 'for', 't...</td>\n      <td>['Prevascularized', 'silicon', 'membranes', 'f...</td>\n      <td>['T169', 'T109', 'T073', 'o', 'o', 'T052', 'o'...</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>Seated maximum flexion: An alternative to stan...</td>\n      <td>['C0277814', 'C0806909', 'C0231452', ':', 'An'...</td>\n      <td>['Seated', 'maximum', 'flexion', ':', 'An', 'a...</td>\n      <td>['T033', 'T081', 'T042', 'o', 'o', 'T077', 'o'...</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>The Relationship Between Distance and Post-ope...</td>\n      <td>['The', 'C0439849', 'Between', 'C0012751', 'an...</td>\n      <td>['The', 'Relationship', 'Between', 'Distance',...</td>\n      <td>['o', 'T080', 'o', 'T081', 'o', 'T033', 'T053'...</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>Promoting lifestyle behaviour change and well-...</td>\n      <td>['C0033414', 'C0870811', 'and', 'C0424578', 'i...</td>\n      <td>['Promoting', 'lifestyle behaviour change', 'a...</td>\n      <td>['T052', 'T054', 'o', 'T033', 'o', 'T101', 'o'...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.7.9","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":4}